### ISSUE TITLE
Display WMS legend from GeoServer

### ISSUE THREAD

**Issue by andyb-esdm** [jncc/web-mapper-core#33]
We will naively display the legend from GetLegendGraphic from GeoServer in a popup (or other appropriate way). We're not sure how configurable the legend graphic is in GeoServer.

**UI:** user clicks control next to layer and mapper requests legend graphic and displays in a control e.g. popup

**Comment by andyb-esdm** [comment]
@JamesPe 
http://jnccdev-geo.esdm.co.uk/emodnet/wms?REQUEST=GetLegendGraphic&VERSION=1.3.0&FORMAT=image/png&WIDTH=25&HEIGHT=25&LAYER=eusm_msfd
Gives me a legend for MSFD at all resolutions in the group i.e. it repeats itself several times
I can specify a single resolution layer like this:
http://jnccdev-geo.esdm.co.uk/emodnet/wms?REQUEST=GetLegendGraphic&VERSION=1.3.0&FORMAT=image/png&WIDTH=25&HEIGHT=25&LAYER=eusm2016_msfd_800
But difficult to see how the mapper can programatically determine the legend to use from the group layer.
Are these resolution dependent layers going to be kept in a group layer? Or will caching mean we won't need to use them in future?

**Comment by JamesPe** [comment]
Not sure that's been decided yet need to do some tests with them.  In hoping we can get rid of multiple layers.  Failing that I guess we need to extend API to include layer for legend.

I think we can assume legends are the same for all layers in a group.

Unless we can ask for legend 1 but never seen that

On 17 Feb 2019 17:08, Andy Brewer <notifications@github.com> wrote:

@JamesPe<https://github.com/JamesPe>
http://jnccdev-geo.esdm.co.uk/emodnet/wms?REQUEST=GetLegendGraphic&VERSION=1.3.0&FORMAT=image/png&WIDTH=25&HEIGHT=25&LAYER=eusm_msfd
Gives me a legend for MSFD at all resolutions in the group i.e. it repeats itself several times
I can specify a single resolution layer like this:
http://jnccdev-geo.esdm.co.uk/emodnet/wms?REQUEST=GetLegendGraphic&VERSION=1.3.0&FORMAT=image/png&WIDTH=25&HEIGHT=25&LAYER=eusm2016_msfd_800
But difficult to see how the mapper can programatically determine the legend to use from the group layer.
Are these resolution dependent layers going to be kept in a group layer? Or will caching mean we won't need to use them in future?

**Comment by andyb-esdm** [comment]
Thanks. Good idea. A legend layer configuration property makes sense if the grouped layers need to remain.

**Comment by andyb-esdm** [comment]
@SimonAnnetts should you add a nullable column LegendLayer (or similar) to the layer configuration table and api?
If populated, this would be used in the GetLegendGraphic request. If null, the mapper client would fall back to the layer's layerName property.

**Comment by SimonAnnetts** [comment]
This sounds like a sensible solution and I will implement it as you proposed.

**Comment by andyb-esdm** [comment]
@SimonAnnetts - thanks - I've already added a property legendLayerName to the ILayerConfig interface in the mapper. So it should just naturally hydrate if data is received.

**Comment by SimonAnnetts** [comment]
Rolled out to our staging server, but their staging server seems to be broken this morning (it can't connect to the database). I have added a reference to one of the layers in a layer group into the test layer data.

**Comment by SimonAnnetts** [comment]
documentation updated:
https://github.com/jncc/web-mapper-core/wiki/Database-Schema-and-API-Endpoints

**Comment by andyb-esdm** [comment]
This is implemented i.e. functional but needs styling/layout applied.

**Comment by JordanPinder** [comment]
@andyb-esdm @SimonAnnetts implemented some real layers and groups on the [staging mapper](https://prelive.mapper.emodnet-seabedhabitats.eu/), but getting an issue with the getLegend requests - request just seems to hang and not pull through. Do you get a similar problem?

**Comment by HelenwoodsJNCC** [comment]
@andyb-esdm @SimonAnnetts seems to be associated with the following error message:
Cross-Origin Read Blocking (CORB) blocked cross-origin response https://staging.ows.emodnet-seabedhabitats.eu/emodnet/wms?REQUEST=GetLegendGraphic&VERSION=1.3.0&FORMAT=image/png&WIDTH=20&HEIGHT=20&LEGEND_OPTIONS=fontAntiAliasing:true;fontColor:0x5A5A5A&LAYER=eusm_all with MIME type text/xml. See https://www.chromestatus.com/feature/5629709824032768 for more details.

**Comment by JordanPinder** [comment]
@andyb-esdm @SimonAnnetts ignore this - I misinterpreted the "LegendLayerName" column, I thought this was in reference to the style name itself, not the layer name. 

Nothing wrong with this, seems to work fine :)

**Comment by HelenwoodsJNCC** [comment]
@SimonAnnetts @andyb-esdm we have had a bit of feedback about the legend window.  Firstly if you already have a GetFeatureInfo request open and then click on the get legend option it doesn't seem to do anything. I think the legend is actually appearing behind but this isn't clear, and the most recently clicked should ideally move to the front. There was also a comment asking if the legend can be filtered when a layer is filtered - but is this even possible with the GetLegend requests?

**Comment by andyb-esdm** [comment]
@HelenwoodsJNCC GetFeatureInfo now closes when you open the legend

**Comment by HelenwoodsJNCC** [comment]
@andyb-esdm thanks! I see the filter box now closes when you click off it too.

**Comment by andyb-esdm** [comment]
@HelenwoodsJNCC Yes. Trying to tidy up remaining loose ends.

**Comment by andyb-esdm** [comment]
@HelenwoodsJNCC I can't find any more appropriate issues for the remaining changes so here they are!

- I've fixed the sort order on the layer groups. They should now respect your configuration. External layer groups will be added to the end of the layer selector.
- The map should now also respect max zoom set on the map instance in the configuration.

**Comment by HelenwoodsJNCC** [comment]
Great, thanks @andyb-esdm

**Comment by HelenwoodsJNCC** [comment]
@andyb-esdm so although the Layer group order has been resolved there still seems to be a bit of an issue with the Layer Order. Doesn't seem to matter what I change the numbers to in the Layer API table they always draw in the same order.

**Comment by andyb-esdm** [comment]
@HelenwoodsJNCC I'm sorry, I didn't know there was a layer order too!
Sublayergroups complicate this a bit because they don't have an order against them (sublayergroup is just a property of the layers that can be in any order).
So, I've tried this:

1. Order the sublayer groups by the lowest order value of its constituent layers e.g sublayergroup1 with layer orders: 2, 202, 303 would come before sublayergroup2 with layer orders: 3, 4, 5
2. Then sort the layers within the sublayergroup by their orders

I've implemented this but need to test it on our infrastructure by changing layer orders.
Also, there's now an embargo on new versions of the mapper for emodnet see #77 
I'll deploy to our test server but let me know if you want the changes to the MPA pre-live environment.
If you do, we'll need to remember to get everything in sync again after the embargo.
